driver-core/src/main/com/mongodb/connection/MultiServerCluster.java

```diff
@@ -375,13 +376,15 @@ private void addNewHosts(final Set<String> hosts) {
 
     private void removeExtraHosts(final ServerDescription serverDescription) {
         Set<ServerAddress> allServerAddresses = getAllServerAddresses(serverDescription);
-        for (final ServerTuple cur : addressToServerTupleMap.values()) {
+        for (Iterator<ServerTuple> iterator = addressToServerTupleMap.values().iterator(); iterator.hasNext();) {
+            ServerTuple cur = iterator.next();
             if (!allServerAddresses.contains(cur.description.getAddress())) {
                 if (LOGGER.isInfoEnabled()) {
                     LOGGER.info(format("Server %s is no longer a member of the replica set.  Removing from client view of cluster.",
                                        cur.description.getAddress()));
                 }
-                removeServer(cur.description.getAddress());
+                iterator.remove();
```

Comment:
This may not be a real problem for a ConcurrentHashMap, but it's more idiomatic to remove via the Iterator.

Reply:
Ack